### ISSUE TITLE
Any idea why source kit is throwing this error - ASTInvocation problem ?

### ISSUE THREAD

**Issue by pkasson** [jpsim/SourceKitten#697]
sourcekit: [1:getCursorInfo: 0.0000] failed to create an ASTInvocation: '/home/ubuntu/dev/swift/SourceKittenUtil/temp/AnishaApril26/AnishaApril26-master/AnishaProduct/ContentView.swift' is not part of the input files

Have seen this on a number of files, and the files exist and appear to be valid structs, or class files.

**Comment by johnfairh** [comment]
Just what the message says really: it means Swift thinks that the source file you are querying (`key.sourcefile`) does not appear in the list of files passed to the compiler for compilation (`key.compilerargs`).

If you think your keys are consistent then you should take a full example to bugs.swift.org.

**Comment by pkasson** [comment]
So, this is not something being compiled, but parsed by source kit - it feels like something wrong with the source, something it does not understand, like SwiftUI or something new it cant comprehend.

**Comment by pkasson** [comment]
How can you catch this error during this parsing process

**Comment by johnfairh** [comment]
No, nothing at all to do with "the parsing process", it's about the keys in your sourcekit request per my previous comment.

**Comment by pkasson** [comment]
So, when source kit starts to analyze the source in a project, it encounters a file it did not get passed for compilation.  

Not sure I fully understand how source kit is doing its thing, but if it is acting on files in a project, using SourceKitten, is it compiling the source in order to obtain the key and structure data ?

**Comment by johnfairh** [comment]
The compiler args are supplied by the user or taken from the build tool CLI/record.

If you are not making sourcekit requests directly, can you explain exactly which sourcekitten interface you are using when you see this problem?

If you are running `sourcekitten doc` or the API equivalent against an Xcode project/SPM package then we'll need to see an example of a project with the problem.

**Comment by pkasson** [comment]
` /*
  * get stucture for a source file
  */
 func getStructure(filePath :String) -> Structure?
 {
  var structure : Structure
  
  do                                                                
  {
   structure = try Structure(file: File(path: filePath)!)
   
   return structure
  }
  catch let error
  {
   logger.logError(message: "getStructure - error:  \(error.localizedDescription)")
  }
  
  return nil
 }`

**Comment by johnfairh** [comment]
None of that sends a cursorInfo request.  If you have a source file that causes the cursorInfo error from the OP when you run that `Structure` query (sends `source.request.editor.open`) then that should go to bugs.swift.org.

**Comment by pkasson** [comment]
Maybe I dont understand what SourceKitten is doing - when it does interact with SourceKit, is it in fact compiling things, and that this is necessary to obtain structure and other artifacts ?

If not, I see that in SourceKit itself, it is indeed getting this error here:

Utils.cpp - gets initCompilerInvocation invoked, which in turn calls resolveSymbolicLinksInInputs

resolveSymbolicLinksInInputs is the method throwing that error:  OS << "'" << PrimaryFile << "' is not part of the input files";

So ... if SourceKitten is calling SourceKit to do its thing, that sounds like a bug in SourceKitten ?

**Comment by johnfairh** [comment]
SourceKit runs various phases of the compiler to implement the requests you send through SourceKitten.

Your OP suggests SourceKit was sent a CursorInfo request.

Your code could be sending this directly, or using SourceKitten's docs generation APIs (`SwiftDocs`, `Module.docs`).
If you are sending a `Request` directly, or using `SwiftDocs` directly, or using `Module.init(name:compilerArguments:)`, then the compiler arguments are under your control.

If you are using SourceKitten's end-to-end docs generation via a different `Module` initializer then SourceKitten is responsible for figuring out the compiler arguments via a build tool.

**Comment by pkasson** [comment]
I took out the call to get documentation .. and I have a lot fewer errors as shown earlier.  
Here is my call, using kitten:

`
 /*
  * get comments / docs for a source file
  */
 func getDoc(filePath :String) -> [String: SourceKitRepresentable]
 {
  var args = [String]()
  args.append("single-file")
  
  let docs = SwiftDocs(file: File(path: filePath)!, arguments: args)
  
  return docs!.docsDictionary
 }
`

This used to work without issue (not all issues as described earlier), but now, some, not all, shoot out tons of error messages about ASTInvocation.  Is this call not correct ?

**Comment by johnfairh** [comment]
The `arguments` here should be the Swift compiler arguments required to compile the module/part-module that contains `filePath`.  The [sourcekit docs](https://github.com/apple/swift/blob/main/tools/SourceKit/docs/Protocol.md#cursor-info) describe this:
>Array of zero or more strings for the compiler arguments, e.g ["-sdk", "/path/to/sdk"]. If key.sourcefile is provided, these must include the path to that file.

SourceKitten passes them through to SourceKit without touching them.
